File: python/janus/chat/ui/display.py
"""Rich-based terminal UI for chat display."""

from rich.console import Console
from rich.live import Live
from rich.markdown import Markdown
from rich.panel import Panel
from rich.spinner import Spinner
from rich.table import Table

# Model color mapping
MODEL_COLORS = {
    "anthropic": "magenta",
    "openai": "green",
    "gemini": "blue",
    "grok": "red",
}


class ChatDisplay:
    """Handles all terminal rendering for chat."""

    def __init__(self, console: Console | None = None):
        self.console = console or Console()

    def show_welcome(self, models: list[str]) -> None:
        """Display welcome message and available models."""
        self.console.print()
        self.console.print("[bold cyan]━━━ Janus Chat Room ━━━[/bold cyan]")
        self.console.print(f"[dim]Models: {', '.join(models)}[/dim]")
        self.console.print("[dim]Type /help for commands, /quit to exit[/dim]")
        self.console.print()

    def show_user_message(self, content: str) -> None:
        """Display user's message."""
        self.console.print(
            Panel(
                content,
                title="[cyan bold]You[/cyan bold]",
                border_style="cyan",
                padding=(0, 1),
            )
        )

    def show_model_response(self, model: str, content: str) -> None:
        """Display a model's response."""
        color = MODEL_COLORS.get(model.lower(), "white")

        # Render as markdown
        rendered = Markdown(content)

        self.console.print(
            Panel(
                rendered,
                title=f"[{color} bold]{model}[/{color} bold]",
                border_style=color,
                padding=(0, 1),
            )
        )

    def show_thinking(self, models: list[str]) -> None:
        """Show thinking indicator for models."""
        names = ", ".join(models)
        self.console.print(f"[dim]  {names} thinking...[/dim]", end="\r")

    def clear_thinking(self) -> None:
        """Clear the thinking indicator."""
        self.console.print(" " * 60, end="\r")

    def show_command_result(self, message: str) -> None:
        """Display command execution result."""
        self.console.print(f"[yellow]{message}[/yellow]")

    def show_error(self, error: str) -> None:
        """Display error message."""
        self.console.print(f"[red bold]Error:[/red bold] {error}")

    def show_help(self) -> None:
        """Display help message."""
        help_text = """
[bold]Commands:[/bold]
  [cyan]/help[/cyan]              Show this help
  [cyan]/quit[/cyan] or [cyan]/exit[/cyan]   Exit chat
  [cyan]/clear[/cyan]             Clear conversation history
  [cyan]/models[/cyan]            Show active models
  [cyan]/expand[/cyan]            Get detailed response to last message
  [cyan]/save[/cyan] [file]       Save session to file
  [cyan]/load[/cyan] <file>       Load previous session

[bold]Tips:[/bold]
  - @anthropic, @openai etc. addresses specific models
  - Responses are kept brief (2-4 sentences) by default
  - Use /expand for detailed explanations
"""
        self.console.print(help_text)

    def get_input(self) -> str:
        """Get input from user with styled prompt."""
        try:
            return self.console.input("[cyan bold]You:[/cyan bold] ")
        except EOFError:
            return "/quit"
        except KeyboardInterrupt:
            self.console.print()
            return "/quit"

File: typescript/src/chat/display.ts
/**
 * Terminal display for chat.
 */

import chalk from 'chalk';
import readline from 'readline';
import { printBanner } from '../utils/banner.js';

// Model color mapping
const MODEL_COLORS: Record<string, chalk.Chalk> = {
    anthropic: chalk.magenta,
    openai: chalk.green,
    gemini: chalk.blue,
    grok: chalk.red,
};

function getModelColor(modelName: string): chalk.Chalk {
    const lowerName = modelName.toLowerCase();
    for (const [key, color] of Object.entries(MODEL_COLORS)) {
        if (lowerName.includes(key)) {
            return color;
        }
    }
    return chalk.white;
}

export class ChatDisplay {
    private rl: readline.Interface | null = null;

    showWelcome(models: string[]): void {
        console.log();
        printBanner('Chat Room');
        console.log();
        console.log(chalk.gray(`Models: ${models.join(', ')}`));
        console.log(chalk.gray('Type /help for commands, /quit to exit'));
        console.log();
    }

    showUserMessage(content: string): void {
        const boxWidth = Math.min(process.stdout.columns || 80, 78);
        const border = '─'.repeat(boxWidth - 2);

        console.log(chalk.cyan(`╭${border}╮`));
        console.log(chalk.cyan(`│ ${chalk.bold('You')}`));
        console.log(chalk.cyan(`├${border}┤`));

        // Word wrap content
        const lines = this.wrapText(content, boxWidth - 4);
        for (const line of lines) {
            console.log(chalk.cyan(`│ `) + line);
        }

        console.log(chalk.cyan(`╰${border}╯`));
    }

    showModelResponse(model: string, content: string): void {
        const color = getModelColor(model);
        const boxWidth = Math.min(process.stdout.columns || 80, 78);
        const border = '─'.repeat(boxWidth - 2);

        console.log(color(`╭${border}╮`));
        console.log(color(`│ ${chalk.bold(model)}`));
        console.log(color(`├${border}┤`));

        // Word wrap content
        const lines = this.wrapText(content, boxWidth - 4);
        for (const line of lines) {
            console.log(color(`│ `) + line);
        }

        console.log(color(`╰${border}╯`));
    }

    showThinking(models: string[]): void {
        const names = models.join(', ');
        process.stdout.write(chalk.gray(`  ${names} thinking...`));
    }

    clearThinking(): void {
        process.stdout.write('\r' + ' '.repeat(60) + '\r');
    }

    showCommandResult(message: string): void {
        console.log(chalk.yellow(message));
    }

    showError(error: string): void {
        console.log(chalk.red.bold('Error: ') + error);
    }

    async getInput(): Promise<string> {
        return new Promise((resolve) => {
            this.rl = readline.createInterface({
                input: process.stdin,
                output: process.stdout,
            });

            this.rl.question(chalk.cyan.bold('You: '), (answer) => {
                this.rl?.close();
                this.rl = null;
                resolve(answer);
            });
        });
    }

    close(): void {
        if (this.rl) {
            this.rl.close();
            this.rl = null;
        }
    }

    private wrapText(text: string, maxWidth: number): string[] {
        const words = text.split(' ');
        const lines: string[] = [];
        let currentLine = '';

        for (const word of words) {
            // Handle newlines in text
            if (word.includes('\n')) {
                const parts = word.split('\n');
                for (let i = 0; i < parts.length; i++) {
                    if (i > 0) {
                        lines.push(currentLine);
                        currentLine = '';
                    }
                    if (parts[i]) {
                        if (currentLine.length + parts[i].length + 1 <= maxWidth) {
                            currentLine += (currentLine ? ' ' : '') + parts[i];
                        } else {
                            if (currentLine) lines.push(currentLine);
                            currentLine = parts[i];
                        }
                    }
                }
                continue;
            }

            if (currentLine.length + word.length + 1 <= maxWidth) {
                currentLine += (currentLine ? ' ' : '') + word;
            } else {
                if (currentLine) lines.push(currentLine);
                currentLine = word;
            }
        }

        if (currentLine) {
            lines.push(currentLine);
        }

        return lines.length > 0 ? lines : [''];
    }
}
